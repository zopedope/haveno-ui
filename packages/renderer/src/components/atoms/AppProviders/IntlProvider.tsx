// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { useEffect, useMemo, useState } from "react";
import { IntlProvider as ReacIntlProvider } from "react-intl";
import type { FC } from "react";
import { LangPack, SupportedLocales } from "@src/constants/lang";
import type { LangKeys } from "@src/constants/lang";

const DEFAULT_LOCALE = SupportedLocales.EN;

export const IntlProvider: FC = ({ children }) => {
  const [messages, setMessages] = useState<{ [key in LangKeys]: string }>();
  const locale = useMemo(
    () =>
      navigator.language in SupportedLocales
        ? navigator.language
        : DEFAULT_LOCALE,
    [navigator.language]
  );

  useEffect(() => {
    setMessages(LangPack[locale] ?? LangPack[DEFAULT_LOCALE]);
  }, [locale]);

  return (
    <ReacIntlProvider locale={locale} messages={messages}>
      {children}
    </ReacIntlProvider>
  );
};
