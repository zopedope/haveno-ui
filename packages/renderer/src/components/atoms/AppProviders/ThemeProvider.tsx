// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { MantineProvider, Global } from "@mantine/core";
import type { FC } from "react";
import { themeOverride, globalStyles } from "@src/theme";

export const ThemeProvider: FC = ({ children }) => {
  return (
    <MantineProvider theme={themeOverride}>
      <Global styles={globalStyles} />
      {children}
    </MantineProvider>
  );
};
