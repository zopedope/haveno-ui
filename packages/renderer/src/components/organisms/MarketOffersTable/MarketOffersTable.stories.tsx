// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { MarketOffersTable } from "./MarketOffersTable";

export default {
  title: "molecules/MarketOffersTable",
  component: MarketOffersTable,
} as ComponentMeta<typeof MarketOffersTable>;

const Template: ComponentStory<typeof MarketOffersTable> = () => {
  return <MarketOffersTable data={data} />;
};

export const Default = Template.bind({});

Default.args = {};

const data = [
  {
    price: 123,
    priceCurrency: "EUR",
    priceComparison: 0.12,
    amount: 123123,
    amountCurrency: "EUR",
    cost: 123,
    costCurrency: "USD",
    paymentMethod: "Bitcoin",
    accountAge: 12,
    accountTrades: 1212,
  },
  {
    price: 123,
    priceCurrency: "EUR",
    priceComparison: 0.12,
    amount: 123123,
    amountCurrency: "EUR",
    cost: 123,
    costCurrency: "USD",
    paymentMethod: "Altcoins",
    accountAge: 12,
    accountTrades: 1212,
  },
];
