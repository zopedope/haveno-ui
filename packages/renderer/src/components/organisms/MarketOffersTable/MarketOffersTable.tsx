// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { createStyles } from "@mantine/core";
import { createTable } from "@tanstack/react-table";
import { useIntl } from "react-intl";
import type { MarketOffer } from "./_types";
import {
  MarketOffersPriceCell,
  MarketOffersAmountCell,
  MarketOffersAccountTradesCell,
  MarketOffersCostsCell,
  MarketOffersAccountAgeCell,
  MarketOffersPaymentCell,
} from "./MarketOffersTableCell";
import { Table } from "@molecules/Table";
import { LangKeys } from "@constants/lang";
import { TableVariant } from "@molecules/Table/_types";

const table = createTable().setRowType<MarketOffer>();

interface MarketOffersTableProps {
  data: Array<MarketOffer>;
}

export function MarketOffersTable({ data }: MarketOffersTableProps) {
  const { classes } = useStyles();
  const columns = useMarketOffersColumns();

  return (
    <Table
      table={table}
      columns={columns}
      data={data}
      variant={TableVariant.Primary}
      state={{
        columnVisibility: {
          accountAge: false,
          accountTrades: false,
        },
      }}
      tableWrap={{
        verticalSpacing: "md",
        className: classes.root,
      }}
    />
  );
}

const useStyles = createStyles((theme) => ({
  root: {
    "thead tr th, tbody tr td": {
      "&:first-of-type": {
        paddingLeft: 30,
      },
      "&:last-of-type": {
        paddingRight: 30,
      },
    },
    "tbody tr": {
      td: {
        background: theme.white,
        paddingTop: 22,
        paddingBottom: 22,
      },
      "&:last-of-type": {
        td: {
          borderBottom: `1px solid ${theme.colors.gray[2]}`,
        },
      },
    },
  },
}));

const useMarketOffersColumns = () => {
  const { formatMessage } = useIntl();

  return [
    table.createDataColumn("price", {
      id: "price",
      header: formatMessage({
        id: LangKeys.MarketsOffersColumnPrice,
        defaultMessage: "Price",
      }),
      cell: ({ row }) => <MarketOffersPriceCell row={row.original} />,
      size: 400,
    }),
    table.createDataColumn("amount", {
      id: "amount",
      header: formatMessage({
        id: LangKeys.MarketsOffersColumnAmount,
        defaultMessage: "Amount",
      }),
      cell: ({ row }) => <MarketOffersAmountCell row={row.original} />,
      size: 400,
    }),
    table.createDataColumn("cost", {
      id: "costs",
      header: formatMessage({
        id: LangKeys.MarketsOffersColumnCost,
        defaultMessage: "Costs",
      }),
      cell: ({ row }) => <MarketOffersCostsCell row={row.original} />,
      size: 400,
    }),
    table.createDataColumn("paymentMethod", {
      id: "paymentMethod",
      header: formatMessage({
        id: LangKeys.MarketsOffersColumnPaymentMethod,
        defaultMessage: "Payment Method",
      }),
      cell: ({ row }) => <MarketOffersPaymentCell row={row.original} />,
      size: 400,
    }),
    table.createDataColumn("accountAge", {
      id: "accountAge",
      header: formatMessage({
        id: LangKeys.MarketsOffersColumnAccountAge,
        defaultMessage: "Account Age",
      }),
      cell: ({ row }) => <MarketOffersAccountAgeCell row={row.original} />,
      size: 400,
    }),
    table.createDataColumn("accountTrades", {
      id: "accountTrades",
      header: formatMessage({
        id: LangKeys.MarketsOffersColumnAccountTrades,
        defaultMessage: "Account Trades",
      }),
      cell: ({ row }) => <MarketOffersAccountTradesCell row={row.original} />,
      size: 400,
      meta: {
        textAlign: "right",
      },
    }),
  ];
};
