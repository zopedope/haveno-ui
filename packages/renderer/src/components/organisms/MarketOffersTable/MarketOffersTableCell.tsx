// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { Group, useMantineTheme } from "@mantine/core";
import type { MarketOffer } from "./_types";
import { Currency } from "@atoms/Currency";
import { BodyText } from "@atoms/Typography";
import { fractionToPercent } from "@src/utils/math";
import { ReactComponent as CheckCircle } from "@assets/check-circle.svg";

export function MarketOffersAccountAgeCell({ row }: { row?: MarketOffer }) {
  const theme = useMantineTheme();

  return (
    <Group spacing="sm">
      <CheckCircle color={theme.colors.blue[6]} width={15} height={15} />
      <BodyText heavy>{row?.accountAge} Days</BodyText>
    </Group>
  );
}

export function MarketOffersPriceCell({ row }: { row?: MarketOffer }) {
  return (
    <Group spacing="sm">
      <BodyText heavy>
        <Currency
          currencyCode={row?.priceCurrency}
          value={row?.price || 0}
          format="symbol"
        />
      </BodyText>

      <BodyText color="gray">
        <Currency
          value={fractionToPercent(row?.priceComparison || 0)}
          minimumFractionDigits={0}
        />
        %
      </BodyText>
    </Group>
  );
}

export function MarketOffersAmountCell({ row }: { row?: MarketOffer }) {
  return (
    <BodyText heavy>
      <Currency currencyCode={row?.amountCurrency} value={row?.amount || 0} />
    </BodyText>
  );
}

export function MarketOffersCostsCell({ row }: { row?: MarketOffer }) {
  return (
    <BodyText heavy>
      <Currency
        currencyCode={row?.costCurrency}
        value={row?.cost || 0}
        format="symbol"
      />
    </BodyText>
  );
}

export function MarketOffersAccountTradesCell({ row }: { row?: MarketOffer }) {
  return (
    <BodyText heavy>
      <Currency value={row?.accountTrades || 0} minimumFractionDigits={0} />
    </BodyText>
  );
}

export function MarketOffersPaymentCell({ row }: { row?: MarketOffer }) {
  return <BodyText heavy>{row?.paymentMethod}</BodyText>;
}
