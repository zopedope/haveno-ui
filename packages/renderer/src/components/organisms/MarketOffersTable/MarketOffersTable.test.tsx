// =============================================================================
//  Copyright 2022 Haveno
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// =============================================================================

import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { MarketOffersTable } from "./MarketOffersTable";
import { AppProviders } from "@atoms/AppProviders";

describe("molecules::MarketOffersTable", () => {
  it("renders without exploding.", () => {
    const { asFragment, unmount } = render(
      <AppProviders>
        <MarketOffersTable data={data} />
      </AppProviders>
    );
    expect(asFragment()).toMatchSnapshot();
    unmount();
  });

  it("renders all columns.", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("Price")).toBeInTheDocument();
    expect(screen.queryByText("Amount")).toBeInTheDocument();
    expect(screen.queryByText("Costs")).toBeInTheDocument();
    expect(screen.queryByText("Payment Method")).toBeInTheDocument();
    unmount();
  });

  it("renders formatted price value with currency sign.", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("€5,000.96")).toBeInTheDocument();
    expect(screen.queryByText("€9,637.41")).toBeInTheDocument();
    unmount();
  });

  it("renders formatted amount value with currency code.", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("XMR 7,564.94")).toBeInTheDocument();
    expect(screen.queryByText("XMR 6,483.23")).toBeInTheDocument();
    unmount();
  });

  it("renders offer formatted payment method.", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("Bitcoin")).toBeInTheDocument();
    expect(screen.queryByText("Altcoins")).toBeInTheDocument();
    unmount();
  });

  it("renders formatted price percentage.", () => {
    const { unmount } = render(
      <AppProviders>
        <MarketOffersTable data={data} />
      </AppProviders>
    );
    expect(screen.queryByText("12%")).toBeInTheDocument();
    expect(screen.queryByText("15%")).toBeInTheDocument();
    unmount();
  });
});

const data = [
  {
    price: 5000.956,
    priceComparison: 0.12,
    priceCurrency: "EUR",
    amount: 7564.94,
    amountCurrency: "XMR",
    cost: 532.34,
    costCurrency: "USD",
    paymentMethod: "Bitcoin",
    accountAge: 12,
    accountTrades: 1212,
  },
  {
    price: 9637.41,
    priceComparison: 0.15,
    priceCurrency: "EUR",
    amount: 6483.23,
    amountCurrency: "XMR",
    cost: 983.32,
    costCurrency: "USD",
    paymentMethod: "Altcoins",
    accountAge: 12,
    accountTrades: 3412,
  },
];
