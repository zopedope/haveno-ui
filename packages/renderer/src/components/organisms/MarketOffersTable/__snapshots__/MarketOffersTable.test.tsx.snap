// Vitest Snapshot v1

exports[`molecules::MarketOffersTable > renders without exploding. 1`] = `
<DocumentFragment>
  <table
    class="mantine-Table-root mantine-shky0d"
  >
    <thead>
      <tr>
        <th
          colspan="1"
          style="width: 400px;"
        >
          Price
        </th>
        <th
          colspan="1"
          style="width: 400px;"
        >
          Amount
        </th>
        <th
          colspan="1"
          style="width: 400px;"
        >
          Costs
        </th>
        <th
          colspan="1"
          style="width: 400px;"
        >
          Payment Method
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Group-root mantine-1lumg83"
          >
            <div
              class="mantine-Text-root mantine-Group-child mantine-1yy4n5b"
            >
              €5,000.96
            </div>
            <div
              class="mantine-Text-root mantine-Group-child mantine-fvwhc7"
            >
              12%
            </div>
          </div>
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Text-root mantine-1o3tzv8"
          >
            XMR 7,564.94
          </div>
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Text-root mantine-1o3tzv8"
          >
            $532.34
          </div>
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Text-root mantine-1o3tzv8"
          >
            Bitcoin
          </div>
        </td>
      </tr>
      <tr>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Group-root mantine-1lumg83"
          >
            <div
              class="mantine-Text-root mantine-Group-child mantine-1yy4n5b"
            >
              €9,637.41
            </div>
            <div
              class="mantine-Text-root mantine-Group-child mantine-fvwhc7"
            >
              15%
            </div>
          </div>
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Text-root mantine-1o3tzv8"
          >
            XMR 6,483.23
          </div>
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Text-root mantine-1o3tzv8"
          >
            $983.32
          </div>
        </td>
        <td
          style="width: 400px;"
        >
          <div
            class="mantine-Text-root mantine-1o3tzv8"
          >
            Altcoins
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
